### DISCUSSION TITLE
MapController With OrthographicView?

### DISCUSSION THREAD

**Discussion by jfuehner** [visgl/deck.gl#6090] [Q&A]
Hi! Is it possible to use the MapController with the OrthographicView? I am looking for a way to visualize EPSG:4326 map layers but maintain the MapController capabilities _(i.e., rotate, tilt, etc)_...

**Comment by Pessimistress** [comment]
Use OrbitView.

**Reply by jfuehner** [reply]
@Pessimistress thats very helpful thank you. 

Having never done what I am attempting to do would you have any deck source code or additional examples you could point me to?

**Reply by Pessimistress** [reply]
[OtbitView](https://deck.gl/docs/api-reference/core/orbit-view)
[Views](https://deck.gl/docs/developer-guide/views)

**Reply by jfuehner** [reply]
@Pessimistress thanks!

I gave the OrbitView a try; see my CodePen (https://codepen.io/jfuehner/pen/ZEBNMBd). I'm not 100% sure I have the `target` property correct...

Something is a bit off, at around zoom scale 8 all the tiles go blank even though I can see in the network console tile images are being returned.

Additionally, when adjusting the pitch via mouse the map doesn't seem to adjust quite like the MapView pitch.

**Reply by Pessimistress** [reply]
Your target z must be always 0 to mimic the MapView's behavior:

```js
initialViewState: {
  target: [256, 256, 0],
  zoom: 4,
  ...
},
onViewStateChange: ({viewState}) => {
  viewState.target[2] = 0;
  return viewState;
}
```

**Reply by jfuehner** [reply]
@Pessimistress makes sense, thanks! 

Not sure if you noticed but in the same CodePen I naively tried to place point icon (IconLayer) using the lat/lon in the 4326 CRS thinking they would magically appear in the correct location on the map.

Not so much… Can you provide some direction where I’ve gone wrong with this?

**Reply by Pessimistress** [reply]
You need to look into the documentation of your tile source and figure out how the tile indices map to coordinates.

**Reply by jfuehner** [reply]
@Pessimistress thanks unfortunately, I think I am out of my depths on this one. This is the service definition...

https://services.arcgisonline.com/arcgis/rest/services/ESRI_Imagery_World_2D/MapServer

Could I use math.gl or proj4 to convert the lat/lon to xy coordinates?

**Reply by jfuehner** [reply]
@Pessimistress 

In this [CodePen](https://codepen.io/jfuehner/pen/ZEBNMBd) I tried reprojecting the `iconLayer` geos based on calculating the x,y positions as seen below... However, I the positions/math are still off...

```
  getPosition: d => {
    const {coordinates} = d;
    const [lng, lat, alt] = coordinates;
    const zoom = Math.ceil(VIEW_STATE.zoom);
    const tileSize = 512;
    const dpi = 96;

    const sinLatitude = Math.sin(lat * Math.PI/180);
    const x = ((lng + 180) / 360) * tileSize * Math.pow(2, zoom);
    const y = (0.5 - Math.log((1 + sinLatitude) / (1 - sinLatitude)) / (4 * Math.PI)) * (tileSize/2) * Math.pow(2, zoom);
    
   const newCoords = [x, y];
    console.log('getPosition', {newCoords, zoom})
    return newCoords;
  }
```

**Reply by jfuehner** [reply]
@Pessimistress I also tried playing around with setting the `coordinateOrigin`/`coordinateSystem` (COORDINATE_SYSTEM.METER_OFFSETS) properties on the `IconLayer`. Although I can get the icon/point to move I can't seem to determine how to calculate how to shift the point to the appropriate location.

Any help would be great....

**Reply by Pessimistress** [reply]
Add these to your TileLayer:

```js
  // If you click on the arcgis link above you can see that the level 0 tiles are 180x180 degrees
  tileSize: 180,
  // import {Matrix4} from 'math.gl'
  // modelMatrix = new Matrix4().scale([1, -1, 1]).translate([-180, -90, 0])
  modelMatrix: [
    1,0,0,0,
    0,-1,0,0,
    0,0,1,0,
    -180,90,0,1
  ],
  // Only fetch within the longitude/latitude range
  extent: [0,0,360,180],
```

Then you can render lng/lat coordinates as is.

**Reply by jfuehner** [reply]
@Pessimistress thanks so much!

Not sure I understand why the `TileLayer` `tileSize` is `180`.  I would have assumed the height/width (512) listed in the service descriptor would need to match the `tileSize` property?

![image](https://user-images.githubusercontent.com/1843941/129817870-074a361b-df78-40ab-a8b2-f334fbcfbd3a.png)

I noticed in my [CodePen](https://codepen.io/jfuehner/pen/ZEBNMBd) when I adjust the pitch tiles are not rendering. Would there be something limiting this with the `OrbitView`? 

![image](https://user-images.githubusercontent.com/1843941/129813644-8a724d62-6f0e-4a14-8d8b-a3a770e17684.png)

**Reply by jfuehner** [reply]
@Pessimistress would there be a way to control how the tiles are filled in? When I get down to lower scales they seem to fill in almost sequentially from left to right as seen in the below video-capture...

https://user-images.githubusercontent.com/1843941/129977705-59738407-2ff6-4d77-a8e2-4936412f7e3a.mov

When viewing this OpenLayers [example](https://openlayers.org/en/latest/examples/xyz-esri-4326-512.html) which uses the same tile source the tile blending is much smoother...


https://user-images.githubusercontent.com/1843941/129978622-794501a1-9bec-4f56-a443-a9bd54cfaedb.mov


Wasn't sure if there was a way to control the tile transitions?

**Reply by jfuehner** [reply]
@Pessimistress with the `OrbitView` what would be the way to get the coordinates of the viewable area/bbox? 

Previously, when I was using the `MapView` I was using the `WebMercatorViewport `to do this:

```
function getWebMercatorViewportViewingRectangle(viewState) {
  const {height, rotationOrbit: bearing, rotationX: pitch, target, width, zoom} = viewState;
  const [longitude, latitude] = target;
  const viewPort = new WebMercatorViewport({bearing, height, latitude, longitude, pitch, target, width, zoom});
  const [upperLeftX, upperLeftY] = viewPort.unproject([0,0]);
  const [upperRightX, upperRightY] = viewPort.unproject([width,0]);
  const [lowerRightX, lowerRightY] = viewPort.unproject([width,height]);
  const [lowerLeftX, lowerLeftY] = viewPort.unproject([0,height]);

  const west = parseFloat(Math.min(...[upperLeftX, upperRightX, lowerLeftX, lowerRightY]))
  const south = parseFloat(Math.min(...[upperLeftY, upperRightY, lowerLeftY, lowerRightY]))
  const east = parseFloat(Math.min(...[upperLeftX, upperRightX, lowerLeftX, lowerRightX]))
  const north = parseFloat(Math.min(...[upperLeftY, upperRightY, lowerLeftY, lowerRightY]))

  return [west, south, east, north];
}
```

This doesn't provide the correct unprojected coordinates. I assume due to me using the `OrbitView`. I did also attempt to use the `Viewport` but again with that I couldn't get it to provide the correct unprojected coordinates.

```
function getViewportViewingRectangle(viewState) {
  const {width, height} = viewState;
  const viewPort = new Viewport(viewState);
  const [upperLeftX, upperLeftY] = viewPort.unproject([0,0]);
  const [upperRightX, upperRightY] = viewPort.unproject([width,0]);
  const [lowerRightX, lowerRightY] = viewPort.unproject([width,height]);
  const [lowerLeftX, lowerLeftY] = viewPort.unproject([0,height]);

  const west = parseFloat(Math.min(...[upperLeftX, upperRightX, lowerLeftX, lowerRightY]))
  const south = parseFloat(Math.min(...[upperLeftY, upperRightY, lowerLeftY, lowerRightY]))
  const east = parseFloat(Math.min(...[upperLeftX, upperRightX, lowerLeftX, lowerRightX]))
  const north = parseFloat(Math.min(...[upperLeftY, upperRightY, lowerLeftY, lowerRightY]))

  return [west, south, east, north];
}
```

You can see the output in this CodePen [example](https://codepen.io/jfuehner/pen/ZEBNMBd) written to the console when `viewState` changes

**Reply by Pessimistress** [reply]
> Not sure I understand why the TileLayer tileSize is 180

If you want to render your icon at coordinate [lng, lat] in an OrbitView, then you need to render the tile corners at their corresponding [lng, lat] also. The first z=0 tile covers the lng-lat range from [-180, -90] to [0, 90], so you need the tile bbox to be exactly that.

A side effect of setting tileSize to `180` rather than `512` is that the TileLayer fetches tiles much more aggressively than needed (if the tiles are "smaller" then we need more tiles to fill the viewport). You can adjust this by setting the `zoomOffset` prop of the TileLayer, probably `-1` or `-2`.

To get the current viewport you can use 

```js
deck.getViewports()[0]
```

To create an arbitrary viewport you can use 

```js
const view = new OrbitView();
view.makeViewport({width, height, viewState});
```

However, `orbitViewport.unproject` will not give you what you are expecting. Because OrbitView is not 2.5D like the MapView, it does not unproject to a point that is on the "ground" (z = 0).

Similarly, because the OrbitView does not have any knowledge of its content, it cannot automatically adjust the near/far clipping plane. You need to estimate them yourself and pass them to the view: https://deck.gl/docs/api-reference/core/view#near

I think it's fair to conclude that OrbitView cannot just replace MapView for your use case. To properly support a EPSG:4326 projection mode you will need to implement a new View class. We do have a long-standing backlog feature request to support non-web-Mercator projections: https://github.com/visgl/deck.gl/blob/master/dev-docs/RFCs/proposals/projection-mode-improvements-rfc.md

**Reply by jfuehner** [reply]
@Pessimistress thanks for your insights, really helpful.

Would you have a recommendation for an equation I may be able to use to calculate/estimate near/far? Looking in the source code of the `WebMercatorViewport` I see a call to `getProjectionProjectionParameters` in the constructor which appears to be calculating the near/far values used to generate its `viewMatrix`. Could I use this same function?

Any idea on why the tiles are not rendering when the map is tilted in `OrbitView `? When using `MapView` I do not experience the issue....

![image](https://user-images.githubusercontent.com/1843941/130263980-0ab13c6e-7fa1-4281-b284-3dbc82c6a38d.png)

**Reply by Pessimistress** [reply]
The tiles are only fetched for the region that is clipped by the far plane.

**Reply by jfuehner** [reply]
@Pessimistress I did a quick [read](https://webglfundamentals.org/webgl/lessons/webgl-3d-perspective.html)  of webgl perspective _(near, far, fov)_ to better educate myself.

I played around with the `fovy`, `near`, `far` and `focalDistance` properties of the `OrbitViewport` _(`Viewport`)_ class. WIth the below properties which appear to match the `WebMercatorViewport` defaults the TileLayer seems to be rendering tiles on the `near` plane *somewhat* _(not really)_ correctly. However, changing the `far` property to various high-low values doesn't seem to have much noticeable impact. 

This [CodePen](https://codepen.io/jfuehner/pen/ZEBNMBd) provides an example...

```
    new OrbitView({
      fovy: 50,
      near: 0.1,
      far: 1000,
      focalDistance: 1.5,
      orthographic: false,
      orbitAxis: 'Z',
      ...
```

Seems like the `OrbitView` has gotten me the majority of the way to replacing the `MapView`. Just not sure why the various values I have experimented with the `fovy`, `near`, `far,` etc are not having the desired effect.

Do those properties need to be updated on every `onViewStateChange?`

I also experimented with implementing my own copy of the `WebMercatorViewport` but didn’t have any luck modifying the `getViewMatrix` function used during construction to create a view matrix that worked for EPSG:4326.

I also noticed if I add a `altitude/z` to the `IconLayer` coordinate positions anything greater than zero is not appearing where I would expect. I assume it has something to do with the `OrbitView` visualizing based on pixel coordinates vs geo-coordinates. Would you have a clever way of handling this you might recommend?

Alternatively, if I can’t get a workable solution with deck’s OrbitView would you know how I could synchronize with OpenLayers? 

This [example](https://openlayers.org/en/latest/examples/xyz-esri-4326-512.html) does 4326. Just would need to somehow synchronize the two frameworks viewState’s.

**Reply by jfuehner** [reply]
@Pessimistress I haven't been able to get the `OrbitView` to work comparable to the `MapView` unfortunately. 

With that, I took the OpenLayers [example](https://openlayers.org/en/latest/examples/xyz-esri-4326-512.html) they provide for the EPSG:4326 tile source and tried to overlay deck as seen in this [CodeSandbox example](https://codesandbox.io/s/xyz-esri-4326-512-forked-038l6?file=/main.js). 

Unfortunately, I can't seem to get the two frameworks views to sync up. You will notice when moving the map around the `IconLayer` icons are out of sync with the basemap. Any suggestions on this?

**Reply by amuedespacher** [reply]
Hi @jfuehner, did you have any luck and time progressing on this topic since 2021? I would like to achieve some thing similar (render a deck.gl map in EPSG:2056 ("swiss projection") instead of the web mercator one), but I am struggling to find the right approach.
